Add ParameterMixin for extending parameter functionality

DCEM opened the pull request:
This commit introduces a general-purpose ParameterMixin designed to allow the extension of QCoDeS Parameters. It enables the addition of new features (e.g., on-cache-change callbacks, interdependent updates, and state reset synchronization) across parameter types without requiring per-parameter customization. It does handle docstring updates, enforces naming and allows to declare compatibility.

It replaces #6700.

<!--

Thanks for submitting a pull request against QCoDeS.

To help us effectively merge your pr please consider the following check list.

- [ ] Make sure that the pull request contains a short description of the changes made.
- [ ] If you are submitting a new feature please document it. This can be in the form of inline
      docstrings, an example notebook or restructured text files.
- [ ] Please include automatic tests for the changes made when possible.

Unless your change is a small or trivial fix please add a small changelog entry:

- [ ] Create a file in the docs\changes\newsfragments folder with a short description of the change.

This file should be in the format number.categoryofcontribution. Here the number should either be the number
of the pull request. To get the number of the pull request one must
first the pull request and then subsequently update the number. The category of contribution should be
one of ``breaking``, ``new``, ``improved``, ``new_driver`` ``improved_driver``, ``underthehood``.

If this fixes a known bug reported against QCoDeS:

- [ ] Please include a string in the following form ``closes #xxx`` where ``xxx``` is the number of the bug fixed.

Please have a look at [the contributing guide](https://microsoft.github.io/Qcodes/community/contributing.html)
for more information.

If you are in doubt about any of this please ask and we will be happy to help.

-->

thangleiter commented:
I think this is a great addition. I have written `Parameter` subclasses with similar patterns [multiple](https://github.com/QCoDeS/Qcodes_contrib_drivers/blob/087e5f45e856c6df3aacd4b0aa3f7e194eba5f50/src/qcodes_contrib_drivers/drivers/SwabianInstruments/private/time_tagger.py#L132-L160) [times](https://github.com/qutech-lab/Qcodes_contrib_drivers/blob/50165e83ab459603e0ad1826e9dd7acb39543afd/src/qcodes_contrib_drivers/drivers/Andor/Andor_iDus4xx.py#L384-L403) + several in my measurement code.

DCEM commented:
@microsoft-github-policy-service agree

DCEM commented:
Hi, after some time away, I’ve rebased and cleaned up this PR. All type checks (mypy, pyright) and tests pass on my end now. I also fixed up the test and requirements files.

There is one possible issue: in the mixin implementation, I’m dynamically patching the parameter cache’s _update_with method to add the desired callback. I used # type: ignore[method-assign] for this line. I believe this is the most straightforward way without touching the cache implementation itself, but please let me know if you’d prefer a different approach or if you see any risks with this pattern.

Let me know if there’s anything unclear or if you’d like changes! Sorry for the long wait - looking forward to your feedback!

jenshnielsen commented:
@DCEM Sorry for the lack of reply. How would you feel about adding the mixins to the qcodes.extensions module rather than the parameters module. This will give us a bit of time to prove out the idea and test it before committing to the Mixins.

jenshnielsen commented:
@DCEM I think something like the ParameterMixin is a good idea. I don't have the time to more detailed think about the design and how it would fit in which makes me think that if we add it as an extension then we have some time to actually use it and test it in real world use cases before committing to extending the parameter modules api.

DCEM commented:
Hi @jenshnielsen,

Moving the mixin to extensions makes sense to me. Having it where it’s easy to use, test, and modify seems like a good choice, and it really is a modular parameter "extension". I’m happy to move the code there.

Two questions on my end:

* For placement: should the mixin go as a single file in `qcodes/extensions,` or would you prefer a subdirectory like 
`qcodes/extensions/parameter/` or `qcodes/extensions/parametermixin/`? Or is there a different location you’d recommend?

* On implementation: to avoid touching the parameter cache code directly, I decided to wrap the parameter cache’s _update_with method dynamically (using # type: ignore[method-assign]). Do you see any risks with this approach, or would you prefer a different method?

Thanks again for your feedback!

DCEM commented:
Hi, thanks for your reply.

I’ve moved the code as discussed. Both mypy and pyright show no issues, and I’ve (hopefully) fixed the pre-commit problems and have reread the commit guide in the process.
It has been a while since i have been working on this, and I don't think i have read this version of it, sorry about that. I might still have missed something since it is somewhat extensive.

The documantation is missing right now - I will have a look on how to work with the .rst files.

jenshnielsen commented on src/qcodes/extensions/parameters/parameter_mixin_set_cache_value_on_reset.py:
```diff
@@ -0,0 +1,127 @@
+        super().__init__(*args, callback=self.reset_cache_value, **kwargs)
+
+        if self.cache_value_after_reset is SetCacheValueOnResetParameterMixin._UNSET:
+            message = f"cache_value_after_reset for parameter '{cast('ParameterBase', self).name}' is not set."
```
The suggestion is worse. If this mixin is useed correctly self is always a ParameterBase

jenshnielsen commented on src/qcodes/extensions/parameters/parameter_mixin_set_cache_value_on_reset.py:
```diff
@@ -0,0 +1,127 @@
+        if set_parser is not None:
+            self.get_parser: Callable[..., Any] | None = lambda x: cast(
+                "ParameterBase", self
+            ).cache.get(get_if_invalid=False)
```
@DCEM Feel free to disregard copilot when it does not make sense

jenshnielsen commented:
@DCEM No worries (sorry about the approve and run the permission management is a bit more restrictive than I would like)

jenshnielsen commented on src/qcodes/extensions/parameters/parameter_mixin_interdependent.py:
```diff
@@ -0,0 +1,195 @@
+        """
+        if method is not None and not callable(method):
+            raise TypeError("dependency_update_method must be callable or None.")
+        self._update_method = method  # type: Callable[[], None] | None
```
```suggestion
        self._update_method: Callable[[], None] | None= method
```

And probably need an import of Callable above

I think this is the last thing causing ci to fail

DCEM commented:
> @DCEM No worries (sorry about the approve and run the permission management is a bit more restrictive than I would like)

Thank you - I think it’s good to be cautious. I’ll try to replicate the CI checks across all supported Python versions in the future so these kinds of issues can hopefully be minimized.

I also want to improve the documentation with Sphinx, but I need to figure out a better workflow for that.
